test/cpp/api/inference_mode.cpp

```
@@ -0,0 +1,391 @@
+  which listed below, even though these tensors are created in InferenceMode, they're
+  normal tensors:
+    - tensors which are view of normal tensors.
+  Note output of inplace ops on normal tensors(which is the input itself).
+*/
+bool is_inference_tensor(torch::Tensor& x) {
+  c10::DispatchKeySet ks = x.key_set();
+  bool has_Autograd = ks.has(c10::DispatchKey::AutogradCPU);
+  bool has_InplaceOrView = ks.has(c10::DispatchKey::InplaceOrView);
+  // They must be either both true or false.
+  bool is_inference_tensor = !has_Autograd && !has_InplaceOrView && x.is_leaf();
+  return is_inference_tensor;
+}
+
+void assert_tensor_creation_meta(torch::Tensor& x, torch::autograd::CreationMeta creation_meta) {
+  ASSERT_EQ(static_cast<torch::autograd::DifferentiableViewMeta*>(x.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta(), creation_meta);
+}
+
+void assert_TLS_states(bool inference_mode) {
+  ASSERT_EQ(InferenceMode::is_enabled(), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_keyset_excluded(c10::autograd_dispatch_keyset), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_key_included(c10::DispatchKey::InplaceOrView), !inference_mode);
+}
+
+TEST(InferenceModeTest, TestTLSState) {
+  assert_TLS_states(false);
+  {
+    InferenceMode guard;
+    assert_TLS_states(true);
+    {
+      InferenceMode guard(false);
+      assert_TLS_states(false);
+    }
+    assert_TLS_states(true);
+  }
+  assert_TLS_states(false);
+}
+
+TEST(InferenceModeTest, TestInferenceTensorCreation) {
+  {
+    InferenceMode guard;
+    // New tensor created through constructors are inference tensors.
+    torch::Tensor c = torch::ones({1, 2, 3});
+    ASSERT_FALSE(c.requires_grad());
+    ASSERT_TRUE(is_inference_tensor(c));
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_TRUE(tmp.requires_grad()); // requires_grad is silently ignored when it's an inference tensor.
```

Did we decide there was a reason we wanted to silently ignore, like code reuse? Seems clear it would be saner devex to actually error if you set requires_grad true on an inference tensor.

---

The code got edited above, but just to be super explicit: it's not silently ignored; we DO set `requires_grad=True` (just like in no grad mode); the idea to not error is so that code that initializes parameters can run in inference mode without triggering errors

---

Yea code reuse is the main reason. And another reason is inference tensor can actually have requires_grad=true (and gradients) as shown in the second PR. :D